source/common/stats/tag_producer_impl.cc

```diff
@@ -101,36 +101,41 @@ void TagProducerImpl::forEachExtractorMatching(
 
 std::string TagProducerImpl::produceTags(absl::string_view metric_name, TagVector& tags) const {
   // TODO(jmarantz): Skip the creation of string-based tags, creating a StatNameTagVector instead.
-  tags.insert(tags.end(), default_tags_.begin(), default_tags_.end());
+  // tags.insert(tags.end(), default_tags_.begin(), default_tags_.end());
   IntervalSetImpl<size_t> remove_characters;
   TagExtractionContext tag_extraction_context(metric_name);
   std::vector<absl::string_view> tokens;
-  forEachExtractorMatching(metric_name, [&remove_characters, &tags, &tag_extraction_context](
-                                            const TagExtractorPtr& tag_extractor) {
-    tag_extractor->extractTag(tag_extraction_context, tags, remove_characters);
+  absl::flat_hash_set<absl::string_view> dup_set;
+  forEachExtractorMatching(metric_name, [&remove_characters, &tags, &tag_extraction_context,
+                                         &dup_set](const TagExtractorPtr& tag_extractor) {
+    bool other_extractor_with_same_name_exists = tag_extractor->otherExtractorWithSameNameExists();
+    if (other_extractor_with_same_name_exists &&
+        dup_set.find(tag_extractor->name()) != dup_set.end()) {
```

Comment:
I think this approach is good. The number of tags on a metric should always be pretty small, so searching them here shouldn't be too expensive. 

This may be faster in practice as a std::vector. IIRC for sizes less than roughly 25 the linear access pattern makes it faster.

Reply:
I'm not sure if find on a vector is better than absl::hash_set, but it's certainly fastest to not check at all when the extractor does not have mate of the same name :)